src/librustc_codegen_llvm/context.rs

```diff
@@ -279,7 +288,24 @@ impl<'a, 'tcx> CodegenCx<'a, 'tcx> {
 
         let isize_ty = Type::ix_llcx(llcx, tcx.data_layout.pointer_size.bits());
 
-        CodegenCx {
+        let alloca_addr_space = tcx.data_layout().alloca_address_space;
+        let mutable_addr_space =
+            tcx.sess.target.target.options.addr_spaces
+              .get(&AddrSpaceKind::ReadWrite)
+              .map(|v| v.index )
+              .unwrap_or_default();
+        let const_addr_space =
+            tcx.sess.target.target.options.addr_spaces
+              .get(&AddrSpaceKind::ReadOnly)
+              .map(|v| v.index )
+              .unwrap_or(mutable_addr_space);
+        let flat_addr_space =
+            tcx.sess.target.target.options.addr_spaces
+              .get(&AddrSpaceKind::Flat)
+              .map(|v| v.index )
+              .unwrap_or_default();
```

Comment:
Is this not always `0`? Is there anything guaranteed about `0`?

Reply:
`0` is always the default (therefore always present), but not *necessarily* the `flat` space. I think a lot of code elsewhere assumes this (and it's true for AMDGPU), but I'm pretty sure it *can* be false.

Reply:
What if you enforce it when parsing the `spec` file? Then we wouldn't just be blindly assuming, and we can keep a lot of code simpler.